components/workspace/FilterContext.tsx

```diff
@@ -0,0 +1,150 @@
+  selectAll: () => void
+  unselectAll: () => void
+  toggleFilteredItems: (select: boolean) => void
+}
+
+const FilterContext = createContext<FilterContextType | null>(null)
+
+const toggleItemCheck = (item: string, select: boolean = true) => {
+  const checkbox = document?.querySelector<HTMLInputElement>(
+    `#mod-workspace input[type="checkbox"][name="workspaceItem"][value="${item}"]`,
+  )
+  if (checkbox) {
+    checkbox.checked = select
+  }
+}
+
+export const FilterProvider = ({
+  children,
+  listData,
+}: {
+  children: React.ReactNode
+  listData?: WorkspaceListData
+}) => {
+  const [filterGroup, setFilterGroup] = useState<FilterGroup[]>([
+    { filters: [] },
+  ])
+
+  const selectAll = () => {
+    if (!listData) return
+    Object.keys(listData).forEach((uri) => {
+      toggleItemCheck(uri)
+    })
+  }
+
+  const unselectAll = () => {
+    if (!listData) return
+    Object.keys(listData).forEach((uri) => {
+      toggleItemCheck(uri, false)
+    })
+  }
+
+  const toggleFilteredItems = (select: boolean) => {
+    if (!listData) return
+    const filteredItems: string[] = []
+
+    for (const [subject, data] of Object.entries(listData)) {
+      let matchesFilters = false
+
+      for (const group of filterGroup) {
+        let groupMatches = false
+
+        for (const filter of group.filters) {
+          if (checkFilterMatchForWorkspaceItem(filter, data)) {
+            groupMatches = true
+          }
+        }
+
+        if (!group.operator) {
+          matchesFilters = groupMatches
+        } else if (group.operator === 'AND' && !groupMatches) {
+          matchesFilters = false
+          break
+        } else if (group.operator === 'OR' && groupMatches) {
+          matchesFilters = true
+          break
+        }
+
+        if (matchesFilters) {
+          filteredItems.push(subject)
+        }
+      }
+    }
+
+    if (filteredItems.length === 0) {
+      toast.error('No items match your filters')
+    } else {
+      for (const filteredItem of filteredItems) {
+        toggleItemCheck(filteredItem, select)
+      }
+    }
+  }
+
+  const addFilter = (groupId: number, filter: WorkspaceFilterItem) => {
+    setFilterGroup((prev) => {
+      const newGroup = [...prev]
+      if (
+        !newGroup[groupId]?.filters.some(
+          (f) => f.field === filter.field && filter.operator === f.operator,
+        )
+      ) {
+        newGroup[groupId].filters.push(filter)
+      }
+      return newGroup
+    })
+  }
+
+  const addGroup = (operator: 'AND' | 'OR') => {
+    setFilterGroup((prev) => [...prev, { operator, filters: [] }])
+  }
+
+  const removeGroup = (groupId: number) => {
+    setFilterGroup((prev) => prev.filter((_, i) => i !== groupId))
+  }
+
+  const removeFilter = (groupId: number, field: string, operator: string) => {
+    setFilterGroup((prev) => {
+      const newGroup = [...prev]
+      newGroup[groupId].filters = newGroup[groupId].filters.filter(
```

Comment:
I think you're cloning `prev` to avoid mutating, but `prev[groupId]` is getting mutated, appearing both in `prev` and `newGroup`.

Reply:
ohh nice catch!